datafusion/functions/src/core/named_struct.rs

```diff
@@ -0,0 +1,186 @@
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow::array::StructArray;
+use arrow::datatypes::{DataType, Field, Fields};
+use datafusion_common::{exec_err, Result, ScalarValue};
+use datafusion_expr::{ColumnarValue, Expr, ExprSchemable};
+use datafusion_expr::{ScalarUDFImpl, Signature, Volatility};
+use std::any::Any;
+use std::sync::Arc;
+
+/// put values in a struct array.
+fn named_struct_expr(args: &[ColumnarValue]) -> Result<ColumnarValue> {
+    // do not accept 0 arguments.
+    if args.is_empty() {
+        return exec_err!("named_struct requires at least one pair of arguments");
+    }
+
+    if args.len() % 2 != 0 {
+        return exec_err!("named_struct requires an even number of arguments");
+    }
+
+    let fields = args
+        .chunks_exact(2)
+        .map(|chunk| {
+            let name = &chunk[0];
+            let value = &chunk[1];
+
+            if let ColumnarValue::Scalar(ScalarValue::Utf8(Some(name))) = name {
+                let array_ref = match value {
+                    ColumnarValue::Array(array) => array.clone(),
+                    ColumnarValue::Scalar(scalar) => scalar.to_array()?.clone(),
+                };
+
+                Ok((
+                    Arc::new(Field::new(
+                        name.clone(),
+                        array_ref.data_type().clone(),
+                        true,
+                    )),
+                    array_ref,
+                ))
+            } else {
+                exec_err!("named_struct even arguments must be string literals")
+            }
+        })
+        .collect::<Result<Vec<_>>>()?;
+    Ok(ColumnarValue::Array(Arc::new(StructArray::from(fields))))
+}
+#[derive(Debug)]
+pub(super) struct NamedStructFunc {
+    signature: Signature,
+}
+
+impl NamedStructFunc {
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::variadic_any(Volatility::Immutable),
+        }
+    }
+}
+
+impl ScalarUDFImpl for NamedStructFunc {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+    fn name(&self) -> &str {
+        "named_struct"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, _arg_types: &[DataType]) -> Result<DataType> {
+        Err(datafusion_common::DataFusionError::Internal(
+            "return_type called instead of return_type_from_exprs".into(),
+        ))
```

Comment:
```suggestion
        unreachable!()
```

Would unreachable!() be suitable here? @alamb @gstvg

Reply:
I can't quite figure out what code this was in relation to (the github UI is being weird). But in general I am a fan of using internal errors if possible even if they really are unreachable. The rationale being that if there is some bug that gets introduced in some future refactoring, then the failure mode is nicer than panic'ing